spec/consensus/light-client/detection.md

```diff
@@ -1,3 +1,459 @@
+and the secondaries are correct, then the failure detector should
+never output evidence for height *h* and should not stop at height *h*.
+
+
+#### **[LCD-VC-INV-DONT-STOP]**
+
+If there is no fork at height *h*, and
+ the primary is correct, then the failure detector should never stop
+ at height *h*.
+
+
+#### **[LCD-VC-LIFE-DONT-STOP]**
+
+If there is no fork at height *h*, and
+ the primary is correct, then the failure detector should eventually
+ decide to not stop and to not report evidence
+ at height *h*.
+
+
+#### **[LCD-VC-LIFE-FORK]**
+
+If there is a fork (two correct full nodes decided on different blocks for the same height), and
+- the light client needs to obtain a header of a height *h* that is
+ affected,
+ and
+- there are two correct full nodes *i* and *j* that are
+    - on different branches, and
+    - primary or secondary,
+
+then the failure detector eventually outputs evidence for height *h*.
+
+
+
+#### **[LCD-REQ-REP]**
+
+If the failure detector observes two conflicting headers for height *h*, it should try to verify both. If both are verified it should report evidence.
+
+If the primary reports header *h* and a secondary reports header *h'*,
+     and if *h'* can be verified based on common root of trust, then
+     evidence should be generated; TODO: shall we add "otherwise drop
+     h' and continue normal operation"?
+
+
+## Definitions
+
+- A fixed set of full nodes is provided in the configuration upon
+     initialization. Initially this set is partitioned into
+    -  one full node that is the *primary* (singleton set),
+	-  a set *Secondaries* (of fixed size, e.g., 3),
+	-  a set *FullNodes*.
+- A set *FaultyNodes* of nodes that the light client suspects of being faulty; it is initially empty
+- *State*  is a set of pairs *(fn,h)* where header *h* has been received (and possibly verified with) full node *fn*.
+- The verifier communicates with the primary [[verification]]. Whenever the verifier successfully verifies a header *h* from the primary *p*, it stores
+ *(p,h)*
+ in *State*.
+
+
+## Solution
+
+
+
+### Inter Process Communication
+
+
+
+For the purpose of this light client specification, we assume that the
+     Tendermint Full Node exposes the following functions over
+     Tendermint RPC:
+
+
+```go
+func Commit(addr Address, height int64) (SignedHeader, error)
+```
+
+- Implementation remark
+   - RPC to full node *n* at address *addr*
+- Expected precodnition
+  - header of `height` exists on blockchain
+- Expected postcondition
+  - if *n* is correct: Returns a sound signed header of height `height`
+  from the blockchain if communication is timely (no timeout)
+  - if *n* is faulty: Returns a signed header with arbitrary content
+- Error condition
+   * if *n* is correct: precondition violated or timeout
+   * if *n* is faulty: arbitrary error
+
+#### **[FN-LuckyCase]**:
+The  full node on which the procedure is called remotely is correct and no timeout occurs at the receiver on `Commit`.
```

Comment:
We can be more specific: instead of "the receiver" we can say the "caller (fork detection mechanism)". Is this what we mean?

Reply:
Yes. The complete truth is that I would like to have the definitions of the RPCs and [FN-LuckyCase] in a separate specification that collects the interface of full nodes for diverse protocols. Anyways, "caller" is definitely more precise!